### ISSUE TITLE
Stack Overflow because of Infinite Loop in Process Directive Evaluation

### ISSUE THREAD

**Issue by carlwitt** [nextflow-io/nextflow#1090]
### Expected behavior and actual behavior

Pipeline terminates with `ERROR ~ Execution aborted due to an unexpected error`
The log files show an infinite recursion during process directive evaluation.

### Steps to reproduce the problem

The problematic portion of the script is where errorStrategy tries to access task.maxRetries and gets stuck in an infinite loop.
```
process metaspades {
    
    // 143 indicates a slurm out of resources error.
    // when this configuration fails after maxRetries, we do not want the whole pipeline to stop 
    errorStrategy { task.exitStatus == 143 ? 'retry' : (task.attempt == task.maxRetries ? 'ignore' : 'retry') }
    maxRetries 5
//   ...
```
Whole workflow (with txt suffix, .nf can't be uploaded)
[benchmark.txt](https://github.com/nextflow-io/nextflow/files/3019329/benchmark.txt)
Test input file (with txt suffix, .csv can't be uploaded)
[benchmark-configurations.txt](https://github.com/nextflow-io/nextflow/files/3019334/benchmark-configurations.txt)

run `nextflow benchmark.nf --configurations benchmark-configurations.csv`
### Program output 

Problem is that GeneratedMethodAccessor384 calls GeneratedMethodAccessor383 and that calls GeneratedMethodAccessor384.
```
java.lang.StackOverflowError: null
        // ... full log file below
        at nextflow.processor.TaskConfig.get(TaskConfig.groovy:108)
        at nextflow.processor.TaskConfig.getErrorStrategy(TaskConfig.groovy:168)
        at nextflow.processor.TaskConfig.getMaxRetries(TaskConfig.groovy:243)
// GeneratedMethodAccessor A
        at sun.reflect.GeneratedMethodAccessor384.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
        at groovy.lang.MetaBeanProperty.getProperty(MetaBeanProperty.java:59)
        at nextflow.processor.TaskConfig.getProperty(TaskConfig.groovy:89)
        at org.codehaus.groovy.runtime.callsite.PogoGetPropertySite.getProperty(PogoGetPropertySite.java:49)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callGetProperty(AbstractCallSite.java:298)
        at _nf_script_63ac27f6$_run_closure4$_closure15.doCall(_nf_script_63ac27f6:148)
        at _nf_script_63ac27f6$_run_closure4$_closure15.doCall(_nf_script_63ac27f6)
// call to GeneratedMethodAccessor B
        at sun.reflect.GeneratedMethodAccessor383.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:104)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:326)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:264)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1041)
        at groovy.lang.Closure.call(Closure.java:411)
        at groovy.lang.Closure.call(Closure.java:405)
        at nextflow.processor.LazyMap.resolveImpl(TaskConfig.groovy:476)
        at nextflow.processor.LazyMap.resolveImpl(TaskConfig.groovy)
        at nextflow.processor.LazyMap.resolve(TaskConfig.groovy:437)
        at nextflow.processor.LazyMap.get(TaskConfig.groovy:500)
        at nextflow.processor.TaskConfig.get(TaskConfig.groovy:108)
        at nextflow.processor.TaskConfig.getErrorStrategy(TaskConfig.groovy:168)
        at nextflow.processor.TaskConfig.getMaxRetries(TaskConfig.groovy:243)
// Call to GeneratedMethodAccessor A here's the cycle
        at sun.reflect.GeneratedMethodAccessor384.invoke(Unknown Source)
```
[nextflow.log](https://github.com/nextflow-io/nextflow/files/3019296/nextflow.log)

### Environment 

* Nextflow version: 19.01.0 
* Java version: 1.8.0_40
* Operating system: macOS

### Additional context

(Add any other context about the problem here)

**Comment by pditommaso** [comment]
Are you able to provide a minimal snippet to replicate the problem? that would help a lot. In any case, I guess that using a separate variable for the max retries value would solve the problem e.g. 


```nextflow
MAX = 5

process metaspades {
    
    // 143 indicates a slurm out of resources error.
    // when this configuration fails after maxRetries, we do not want the whole pipeline to stop 
    errorStrategy { task.exitStatus == 143 ? 'retry' : (task.attempt == MAX ? 'ignore' : 'retry') }
    maxRetries MAX

    /your_command_here/
}
```

**Comment by carlwitt** [comment]
I tried to replicate the problem on my laptop, but I couldn't. I'll try to replicate the problem on the servers I'm using.

**Comment by rsuchecki** [comment]
TLDR; Use `process.maxRetries` not `task.maxRetries`

Note tu future self (and others who stumble here) when setting up a dynamic error strategy (in a config file).

```groovy
process {
  maxRetries = 4
  withName: metaspades {
    errorStrategy = { task.exitStatus == 143 ? 'retry' : (task.attempt == process.maxRetries ? 'ignore' : 'retry') }
  }
}
```
